iceoryx2/src/node/mod.rs

```diff
@@ -408,30 +492,55 @@ impl<Service: service::Service> Node<Service> {
     }
 
     /// Returns a list of [`NodeState`] of all [`Node`]s in the system under a given [`Config`].
-    pub fn list(config: &Config) -> Result<Vec<NodeState<Service>>, NodeListFailure> {
+    ///
+    /// ```
+    /// # use iceoryx2::prelude::*;
+    /// Node::<zero_copy::Service>::list(Config::get_global_config(), |node_state| {
+    ///     println!("found node {:?}", node_state?);
+    ///     Ok(CallbackProgression::Continue)
+    /// });
+    /// ```
+    pub fn list<
+        F: FnMut(
+            Result<NodeState<Service>, NodeListFailure>,
+        ) -> Result<CallbackProgression, NodeListFailure>,
+    >(
+        config: &Config,
+        mut callback: F,
+    ) -> Result<(), NodeListFailure> {
```

Comment:
I feared that the callback with the result input parameter would be a public API. I think for the C and C++ bindings it would be less challenging to have a different API with two callbacks, one for the happy path and one for the error path. The `Result` as return parameter still makes it challenging but still easier to implement than with the `Result` as input parameter.

Reply:
Since this is an API discussion and not a technical issue, it can be resolved in a follow up PR